module-bsp/board/rt1051/puretx/bsp/keyboard/keyboard.cpp

```diff
@@ -67,101 +77,140 @@ namespace bsp::keyboard
                 }
             }
 
-            gpio->EnableInterrupt(1U << static_cast<uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON));
+            gpio->EnableInterrupt(1U << static_cast<std::uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON));
         }
-
-        void clearAllIRQs()
-        {
-            std::uint8_t clearIRQs = 0xFF;
-            i2cAddr.subAddress     = REG_INT_STAT;
-            i2c->Write(i2cAddr, reinterpret_cast<std::uint8_t *>(&clearIRQs), 1);
-        }
-
     } // anonymous namespace
 
-    std::int32_t init(xQueueHandle qHandle)
+    void power(KeyboardPower state)
     {
+        gpio->WritePin(static_cast<std::uint32_t>(BoardDefinitions::KEYBOARD_RESET_PIN),
+                       static_cast<std::uint8_t>(state));
+    }
 
-        i2c = DriverI2C::Create(
-            static_cast<I2CInstances>(BoardDefinitions::KEYBOARD_I2C),
-            DriverI2CParams{.baudrate = static_cast<uint32_t>(BoardDefinitions::KEYBOARD_I2C_BAUDRATE)});
-
-        gpio = DriverGPIO::Create(static_cast<GPIOInstances>(BoardDefinitions::KEYBOARD_GPIO), DriverGPIOParams{});
-
-        qHandleIrq = qHandle;
+    void reset()
+    {
+        // According to datasheet, section 6.7
+        power(KeyboardPower::off);
+        // Reset pulse should be at least 120us
+        vTaskDelay(pdMS_TO_TICKS(chipResetMs));
+        power(KeyboardPower::on);
+        // Recovery time is 120us
+        vTaskDelay(pdMS_TO_TICKS(chipRecoveryMs));
+    }
 
-        gpio->ConfPin(DriverGPIOPinParams{.dir      = DriverGPIOPinParams::Direction::Input,
-                                          .irqMode  = DriverGPIOPinParams::InterruptMode::IntRisingOrFallingEdge,
-                                          .defLogic = 0,
-                                          .pin      = static_cast<uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON)});
+    bool write(const drivers::I2CAddress &addr, const std::uint8_t *txBuff, const std::size_t size)
+    {
+        auto ret = 0;
+        for (auto i = 0; i < maxTransmissionRetry; ++i) {
+            ret = i2c->Write(addr, txBuff, size);
+            if (ret == size) {
+                return true;
+            }
+            vTaskDelay(pdMS_TO_TICKS(retryMs[i]));
+        }
+        LOG_ERROR("Can't write data to keyboard. I2C error: %d", ret);
+        return false;
+    }
 
-        gpio->ConfPin(DriverGPIOPinParams{.dir      = DriverGPIOPinParams::Direction::Output,
-                                          .irqMode  = DriverGPIOPinParams::InterruptMode::NoIntmode,
-                                          .defLogic = 0,
-                                          .pin      = static_cast<uint32_t>(BoardDefinitions::KEYBOARD_RESET_PIN)});
+    bool read(const drivers::I2CAddress &addr, std::uint8_t *rxBuff, const std::size_t size)
```

Comment:
consider adding [[nodiscard]]

Reply:
As we don't have any logic for handling errors we don't use always return code.

Reply:
fair enough